docs/proposals/20210628-nodepool-ingress-support.md

```diff
@@ -186,44 +239,105 @@ As an end user, I want to balance the traffic into a NodePool with layer-7 load-
-	  which will trigger the ingress controller to reload nginx config for the resource state sync.
-	  ```yaml
-            nodepool.openyurt.io/ingress-config: "xxx"
-	  ```
+#### 1). Yurt-app-manager:
+
+Implement a standalone `YurtIngress` operator to support the NodePool ingress feature, which means that `YurtIngress`
+together with `NodePool/YurtAppSet(UnitedDeployment)/YurtAppDaemon` will compose the the current OpenYurt unitization
+capability, and they are all managed by Yurt-app-manager. The responsibility of `YurtIngress` is to monitor the update
+of `YurtIngress` CR and trigger the corresponding operations to nginx ingress controller related components.
+
+`YurtIngress` CRD definition:
+
+```go
+// YurtIngressSpec defines the desired state of YurtIngress
+type YurtIngressSpec struct {
+	// Indicates the replicas of the ingress controllers to be deployed under all the specified nodepools.
+	// +optional
+	Replicas int32 `json:"ingress_controller_replicas_per_pool,omitempty"`
+
+	// Indicates all the nodepools on which to enable ingress.
+	// +optional
+	Pools []string `json:"pools,omitempty"`
+}
+
+// YurtIngressStatus defines the observed state of YurtIngress
+type YurtIngressStatus struct {
+	// Indicates the number of the ingress controllers deployed under all the specified nodepools.
+	// +optional
+	Replicas int32 `json:"ingress_controller_replicas_per_pool,omitempty"`
+
+	// Indicates all the nodepools on which to enable ingress.
+	// +optional
+	Pools []string `json:"pools,omitempty"`
+
+	// Indicates the nginx ingress controller version deployed under all the specified nodepools.
+	// +optional
+	Version string `json:"nginx_ingress_controller_version,omitempty"`
+
+	// Total number of ready pools on which ingress is enabled.
+	// +optional
+	ReadyPoolNum int32 `json:"readyPoolNum"`
+
+	// Total number of unready pools on which ingress is enabling or enable failed.
+	// +optional
+	UnreadyPoolNum int32 `json:"unreadyPoolNum"`
+}
+```
+Some other details about the design:
+- The `YurtIngress` CR should be singleton CR which means there should be only one instance in the cluster,
+  otherwise, it may lead to conflict if users create different `YurtIngress` CR for a same NodePool. We can to
+  implement the related admission webhook to control it.
+- In Cloud Edge usage scenarios, admission webhook should be deployed on the Cloud side for Edge network limitation,
+  so when `YurtIngress Controller` tries to deploy nginx ingress controller, the upstream nginx ingress controller
+  will be divided into 2 parts: one part is deployed on the Cloud side focusing on admission webhook, the other part
+  is deployed to NodePool focusing on service proxy and load balancer. These 2 parts can work collaboratively without
+  any conflict by setting the different startup parameters.
+- A new NodePool binding namespace will be created when a NodePool ingress is enabled, then all the nginx ingress
```

Comment:
A new namespace per NodePool? a lot of new namespaces maybe created.

Reply:
Yes, nginx ingress controller will deploy several components (RBAC/deployment/service/configmap/job/...) in the cluster, if all the related components of different NodePools are deployed in one namespace(e.g. kube-system):
1). We need to define the naming rule to differentiate them
2). Even we can differentiate them by different names, it's not very clear to users
So I suggest to create a namespace per pool for all the related components, besides, I think maybe it can be reused by other features(e.g. CoreDNS) in future.

Reply:
@zzguang ok, i got it. we may also discuss this point in the next community meeting.

Reply:
Okay, thanks!